Write the k-modes chapter (#75)
daffidwilde/thesis: chapters/kmodes/main.tex

\myrule%

This chapter differs from the manuscript by including a more detailed
description of the \(k\)-modes algorithm and its setting (in
Section~\ref{sec:intro}) and omits much of the discussion around matching games
(Section~\ref{sec:inits}) which has now been expanded to form parts of
Chapter~\ref{chp:matching}.

\section{Introduction}\label{sec:intro}

The case study at the end of Chapter~\ref{chp:edo} examined the \(k\)-means
algorithm --- a method for clustering a set of numeric data into \(k\) parts
using Euclidean distance and the cluster means. This chapter considers another
algorithm in the \(k\)-means paradigm, \(k\)-modes, described in a set of
seminal papers by Huang~\cite{Huang1997a, Huang1997b, Huang1998}. The
\(k\)-modes algorithm makes use of similar principles to \(k\)-means, i.e.
adjusting the partition toward some Voronoi tessellation according to a measure
of centrality. However, the method allows for the clustering of categorical data

> by considering a cluster's modal values rather than its mean.

Author: Fixed by 0c7a869

The focus of this chapter is on the initialisation of the \(k\)-modes algorithm.
Like \(k\)-means, there is no guarantee that any two runs of \(k\)-modes will
provide the same clusters; this is due to the stochastic nature of its
initialisation. Therefore, the performance of the algorithm is contingent upon

> the quality of this initial solution.

Author: Fixed by 80e2164

assignment problem (HR) to incorporate mathematical fairness to the initial
solution and to eliminate the greedy component of Huang's method.

In addition to Huang's method, the literature revealed that the next most
commonly cited initialisation was presented by Cao in~\cite{Cao2009}. This
initialisation (referred to as Cao's method) forms the basis of many other
initialisations where a notion of density is central. These two initialisations
form the set of established methods against which the proposed shall be
measured.

This chapter concludes with an analysis of all three initialisations and
demonstrates that the proposed method can outperform both of the established
initialisations using the traditional approach with benchmark datasets. In
addition to this, the method introduced in Chapter~\ref{chp:edo} provides a more
in-depth examination of the methods and how they perform against one another.
The first stage of this analysis serves the purpose of highlighting the merits
of each method in a familiar setting, and while the second stage bolsters these
observations, it provides a vigorous defence of the proposed method and exposes

> the scenarios in which it excels.

Author: Fixed by 0c7a869

The chapter is structured as follows:
\begin{itemize}
    \item Section~\ref{sec:intro} introduces the \(k\)-modes algorithm and its
        components.
    \item Section~\ref{sec:inits} provides an overview of the established
        initialisation methods before a statement of the proposed
        initialisation.
    \item Section~\ref{sec:results} presents analyses of the initialisations on
        benchmark and artificial datasets.
    \item Section~\ref{sec:summary} concludes the chapter.
\end{itemize}

\subsection{The \(k\)-modes algorithm}\label{subsec:kmodes}

Although the case study in Chapter~\ref{chp:edo} made use of the \(k\)-means
algorithm, its notation was never formally introduced. Since this chapter

> considers the \(k\)-modes algorithm more directly, its notation is. As such, the

Author: Yeah, I agree. I'll get rid of the waffle now 👍

> considers the \(k\)-modes algorithm more directly, its notation is. As such, the

Author: Fixed by be524c4

        z^{(k)}\right\}\).
\end{itemize}

As has already been mentioned, the \(k\)-modes algorithm relies on another
distance measure to traditional Euclidean distance since it breaks down in a
categorical attribute space; if the space were ordinal, like the natural
numbers, then some sense of Euclidean distance could be recovered. However, in
the absence of direction or what comes between two points, this distance has no
sensible meaning.

There are numerous measures available for the handling of categorical data.
Definition~\ref{def:dissim} describes what is perhaps the simplest of those

> measures, and is the measure used in the seminal works by Huang. The pitfall of

Author: Fixed by 80e2164

> this measure is that it democratises the attribute space. In doing so, it does

Author: Thank you very much, Vince 😄 I was a little worried about that one, I won't lie.

        \end{cases}
    \end{equation}
\end{definition}

> % TODO include proof of being distance metric?

Author: Fixed by 38ac5d3

With the definition of a categorical distance metric, the notion of a
representative point of a cluster can be addressed. When considering numeric
data with \(k\)-means, a representative point is the mean of the points within
the cluster. With categorical data, however, the mode is used as the measure for
central tendency. This change follows from the concept of dissimilarity in that
the point that best represents (i.e.\ is closest to) those in a cluster is one
with the most frequent attribute values of the points in the cluster. The
following definitions formalise this notion, and Theorem~\ref{thm:mode} provides

> a method to find such a point.

Author: I think an example would be nice in this section but I'm not entirely sure what you mean.

            }
        }{N}
        = \frac{n \left(x_j^{(i)}\right)}{N}
    \end{equation}

    That is, \(\text{Dens}_j \left(X^{(i)}\right)\) is the relative
    frequency of \(x_j^{(i)}\) in \(\mathcal X\). Also, observe that:
    \begin{equation}
        \abs*{\left\{X^{(t)} \in \mathcal{X} : x_j^{(i)} = x_j^{(t)}\right\}}%
        = \sum_{t=1}^N \left(1 - \delta\left(x_j^{(i)}, x_j^{(t)}\right)\right)
    \end{equation}

    Hence, an alternative definition for~\eqref{eq:density} can be derived:
    \begin{equation}\label{eq:density-alt}
        \text{Dens}\left(X^{(i)}\right)
        = \frac{1}{mN} \sum_{j=1}^m \sum_{t=1}^N \left(%
            1 - \delta\left(x_j^{(i)}, x_j^{(t)}\right)
        \right)
        = 1 - \frac{1}{mN} D\left(\mathcal{X}, X^{(i)}\right)
    \end{equation}

    This alternative definition highlights how identifying high-density data
    points would mean finding data points with a low summed dissimilarity, i.e.
    true modes of the dataset.
\end{definition}

\inputalg{cao}

\subsection{The proposed method}\label{subsec:method}

Both of the initialisation methods described in this section have a greedy

> component. Cao's method essentially chooses the densest point that has not

Author: I believe it is but I prefer your wording.

this entirely as there are certainly benefits to comparing methods in this way
such as familiarity and direct comparison. However, as that discussion
elaborates on, it does not afford a rich understanding of how any of the methods
perform more generally. This stage of the analysis relies on the method for
generating artificial datasets introduced in Chapter~\ref{chp:edo}, and, as was
made clear in that chapter's case study, the critical component of this method
is a well-defined fitness function.

In order to expose the nuances in the performance of Cao's method and the
proposed initialisation on a particular dataset, two cases are considered: where
Cao's method outperforms the proposed, and vice versa. Both cases use the same
fitness function, although the latter uses its negative. The function is defined
as follows:
\begin{equation}\label{eq:fitness}
    f\left(\mathcal X\right) = C_{\mathrm{cao}} - C_{\mathrm{match}}
\end{equation}

Here, \(C_{\mathrm{cao}}\) and \(C_{\mathrm{match}}\) are the final costs when a
dataset \(\mathcal X\) is clustered using Cao's method and the proposed matching
method respectively with \(k = 3\). Given that costs should be minimised, this

> function is given to EDO to be minimised. As such, the case that uses \(f\) will
> be referred to as Cao-preferable. For the sake of computational time, the

Reviewer: Is this worth having?

```suggestion
function is given to EDO to be minimised. As such, the case that uses \(f\) will
be referred to as Cao-preferable and \(-f\) will be referred to as matching-preferable. For the sake of computational time, the
```

proposed initialisation is given 25 repetitions.

Apart from the sign of \(f\), each case uses identical EDO parameters: a
population size of \(N = 500\); row and column limits of \(R = (50, 500)\) and
\(C = (2, 50)\), respectively; selection parameters \(b = 0.2\) and \(l = 0\); a
mutation probability of \(p_m = 0.01\); a maximum number of iterations of \(M =
100\). In addition to these parameters, the only distribution family included is
a discrete uniform family with up to 10 categories. This means that all columns
in the generated datasets will have at least one category but no more than 10.

\begin{figure}
    \centering
    \includegraphics[width=\imgwidth]{Fig9.pdf}
    \caption{%
        Histograms of fitness for the top performing percentile in each case.
    }\label{fig:fitness}
\end{figure}

> This process yielded approximately 35,000 unique datasets for each case, and the

Reviewer: Can we add a link to the archived data here?

Author: Fixed by 80e2164

is represented as their first principal component obtained via centred Principal
Component Analysis (PCA)~\cite{Jolliffe1986}. While some subtleties may be lost,
this representation captures the essential characteristics of each dataset in a
single variable meaning they can be compared directly.

Since the transformation by PCA is centred, all measures for central tendency
are moot. The mean and median are not interpretable here anyway, given that the
original data is categorical. As such, the univariate statistics used here
describe the spread and shape of the principal components and are split into
two groups: central moments (variance, skewness and kurtosis) and empirical
quantiles (the interquartile range, and lower and upper deciles).

Figures~\ref{fig:edo_moments}~and~\ref{fig:edo_quantiles} show the distributions
of the six univariate statistics across all of the principal components in each
case. In addition to this, they show a fitted Gaussian kernel density
estimate~\cite{Bashtannyk2001} to accentuate the general shape of the
histograms. What is immediately clear from each of these plots is that for

> datasets where Cao's method succeeds, the general spread of their first

Reviewer: ```suggestion
datasets where Cao's method performs better, the general spread of their first
```

\section{Chapter summary}\label{sec:summary}

This chapter has introduced a novel initialisation method for the \(k\)-modes
algorithm that builds on the method set out in the seminal
paper~\cite{Huang1998}; the new method models the final replacement process in
the original as an instance of the hospital-resident assignment problem that may
be solved to be mathematically fair and stable using a Gale-Shapley algorithm.

Following a thorough description of the \(k\)-modes algorithm and the
established initialisation methods, a comparative analysis was conducted among
the three initialisations using benchmark datasets. In addition, the EDO method
described in Chapter~\ref{chp:edo} was utilised to extend this analysis to
artificial datasets.

The first of these analyses revealed that the proposed initialisation was able
to consistently outperform both of the other methods when the choice of \(k\)
had been optimised somewhat. However, the proposed method was unable to beat
Cao's method when an external, and frankly separate, framework was imposed on
each dataset by choosing \(k\) to be the number of classes present.

> The latter analysis showed that proposed method should be employed over Cao's if

Reviewer: ```suggestion
The latter analysis showed that the proposed method should be employed over Cao's if
```
